### PR TITLE
Temporaries for MustCallChecker

### PR DESCRIPTION
**Pull request by Nargeshdb**
We made MustCallChecker aware of temporary variables in this PR.  To do that, we create temporaries and add them into the store in MCC. There are two failing tests after these changes that @kelloggm is working on them.

### CODE REVIEW

File: object-construction-checker/tests/socket/BindChannel.java [hunk 1]
```diff
@@ -15,7 +15,7 @@ static void test(InetSocketAddress addr, boolean b) {
             // it's okay to keep issuing this error; see the second method in this class
             // for an example of how to rewrite the code to avoid the parse error.
             //
-            // :: error: required.method.not.called
+            // TODO we should report a warning here
```
Comment (kelloggm) [inline object-construction-checker/tests/socket/BindChannel.java:18]: This is okay for now, because we report the error below. I'm investigating this.

File: object-construction-checker/tests/socket/HdfsReport1.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Based on a false positive in hdfs
```
Comment (kelloggm) [inline object-construction-checker/tests/socket/HdfsReport1.java:1]: Should we include this unrelated test in this PR? I would think it would go in a separate PR.

File: object-construction-checker/tests/socket/HdfsReport2.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// Based on a false positive in hdfs
```
Comment (kelloggm) [inline object-construction-checker/tests/socket/HdfsReport2.java:1]: Same comment as HdfsReport1.java

File: object-construction-checker/tests/socket/HdfsReport3.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public void dumpTreeRecursively(PrintWriter out, StringBuilder prefix,
+                                    int snapshotId) {
+//        printing something
+    }
+
+    // StringBuilder doesn't implement closeable
+    final private StringBuilder sb = new StringBuilder();
+    final private Formatter formatter = new Formatter(sb);
+
+}
```
Comment (kelloggm) [inline object-construction-checker/tests/socket/HdfsReport3.java:38]: nit: no newline

**Review by kelloggm — APPROVED** [review]

File: object-construction-checker/tests/socket/ZookeeperReport6.java [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
 
-// @skip-test
-
 import java.nio.channels.SocketChannel;
 import java.io.IOException;
 
 class ZookeeperReport6 {
     SocketChannel createSock() throws IOException {
         SocketChannel sock;
+        // false positive
```
Comment (msridhar) [inline object-construction-checker/tests/socket/ZookeeperReport6.java:9]: why change skipping the test to a comment about a false positive?

File: object-construction-checker/tests/mustcall/CommonModuleCrash.java [hunk 6]
```diff
@@ -1,11 +1,9 @@
 import java.net.*;
 
 class CommonModuleCrash {
-    // :: error: required.method.not.called
```
Comment (msridhar) [inline object-construction-checker/tests/mustcall/CommonModuleCrash.java:4]: why this change?

File: object-construction-checker/tests/mustcall/CommonModuleCrash.java [hunk 7]
```diff
@@ -1,11 +1,9 @@
...
     Socket bar = new Socket();
     static void baz(Socket s) {  }
     static {
-        // :: error: required.method.not.called
```
Comment (msridhar) [inline object-construction-checker/tests/mustcall/CommonModuleCrash.java:8]: why this change?

**Review by msridhar — COMMENTED** [review]
Is this pull request complete?  I see code for introducing temp vars in the MCC, but I don't see any code in the MCC that actually uses it.  Is the idea that we would land this and then add actual support in the MCC in a follow-up?

File: object-construction-checker/tests/mustcall/CommonModuleCrash.java (hunk 6 above)
Comment (Nargeshdb) [reply object-construction-checker/tests/mustcall/CommonModuleCrash.java:4]: Because new Socket creates unbound socket which is MCC{}

File: object-construction-checker/tests/mustcall/CommonModuleCrash.java (hunk 7 above)
Comment (kelloggm) [reply object-construction-checker/tests/mustcall/CommonModuleCrash.java:8]: These were false positives - these sockets are unconnected.

Comment (Nargeshdb) [reply object-construction-checker/tests/mustcall/CommonModuleCrash.java:8]: same reason

**Comment by kelloggm** [comment]
@msridhar though there's not any code in the MCC that uses tempvars directly, this change already has significant impact. That's because the MustCallInvokedChecker looks up the type of various nodes in the Must Call Checker's store. With this change, it's able to look up the value of temporary variables - so, for example, it's able to reason that an expression that's MCC with an unconnected socket or `StringReader` is also `@MustCall({})`. Previously, to support resets, it was necessary to assume the worst about expressions that weren't in the MC store; that is, if we didn't know the type of an expression, we had to assume it was reset somewhere and therefore had its class' default type. This change makes that no longer necessary in a lot of cases, improving precision.

**Review by msridhar — APPROVED** [review]

**Comment by kelloggm** [comment]
@msridhar @Nargeshdb I've fixed the false positive in `ZookeeperReport6.java`, and updated the explanation for why `BindChannel.java` issues the error where it does. This PR should now be ready to merge.

File: object-construction-checker/tests/socket/ZookeeperReport6.java (hunk 5 above)
Comment (kelloggm) [reply object-construction-checker/tests/socket/ZookeeperReport6.java]: because I was fixing it, see my latest commit
